File: src/main/java/it/giaquinto/springberry/model/specific/memory/MemoryProperties.java
package it.giaquinto.springberry.model.specific.memory;

import it.giaquinto.springberry.model.security.SecurityLevel;

import java.lang.management.ManagementFactory;
import java.lang.management.MemoryUsage;

public final class MemoryProperties {

    private final ReadableMemoryProperties heap;

    private final ReadableMemoryProperties nonHeap;

    public MemoryProperties(final SecurityLevel level) {
        var memory = ManagementFactory.getMemoryMXBean();

        if (level == SecurityLevel.ADMIN || level == SecurityLevel.POWER) {
            memory.setVerbose(true);
        }

        heap = new ReadableMemoryProperties(memory.getHeapMemoryUsage());

        nonHeap = new ReadableMemoryProperties(memory.getNonHeapMemoryUsage());
    }

    public ReadableMemoryProperties getHeap() {
        return heap;
    }

    public ReadableMemoryProperties getNonHeap() {
        return nonHeap;
    }

    @Override
    public String toString() {
        return "MemoryProperties{" + "heap=" + heap + ", nonHeap=" + nonHeap + '}';
    }
}

File: src/main/java/it/giaquinto/springberry/model/specific/memory/MemoryProperties.kt
package it.giaquinto.springberry.model.specific.memory

import it.giaquinto.springberry.model.security.SecurityLevel
import java.lang.management.ManagementFactory

class MemoryProperties(level: SecurityLevel) {
    val heap: ReadableMemoryProperties
    val nonHeap: ReadableMemoryProperties

    init {
        val memory = ManagementFactory.getMemoryMXBean()
        if (level == SecurityLevel.ADMIN || level == SecurityLevel.POWER) {
            memory.isVerbose = true
        }
        heap = ReadableMemoryProperties(memory.heapMemoryUsage)
        nonHeap = ReadableMemoryProperties(memory.nonHeapMemoryUsage)
    }

    override fun toString(): String {
        return "MemoryProperties{heap=$heap, nonHeap=$nonHeap}"
    }
}